cluster-autoscaler/simulator/dynamicresources/snapshot/patchset_test.go

```diff
@@ -0,0 +1,821 @@
+	}{
+		"CommitEmptyPatchSet": {
+			patchLayers: []map[string]*int{},
+			wantMap:     map[string]int{},
+		},
+		"CommitSingleLayer": {
+			patchLayers: []map[string]*int{{"a": pint(1)}},
+			wantMap:     map[string]int{"a": 1},
+		},
+		"CommitTwoLayersNoOverlap": {
+			patchLayers: []map[string]*int{{"a": pint(1)}, {"b": pint(2)}},
+			wantMap:     map[string]int{"a": 1, "b": 2},
+		},
+		"CommitTwoLayersWithOverwrite": {
+			patchLayers: []map[string]*int{{"a": pint(1)}, {"a": pint(2)}},
+			wantMap:     map[string]int{"a": 2},
+		},
+		"CommitTwoLayersWithDeleteInTopLayer": {
+			patchLayers: []map[string]*int{{"a": pint(1), "b": pint(5)}, {"c": pint(3), "a": nil}},
+			wantMap:     map[string]int{"b": 5, "c": 3},
+		},
+		"CommitTwoLayersWithDeleteInBottomLayer": {
+			patchLayers: []map[string]*int{{"a": pint(1), "b": nil}, {"a": pint(11), "c": pint(3)}},
+			wantMap:     map[string]int{"a": 11, "c": 3},
+		},
+		"CommitThreeLayers": {
+			patchLayers: []map[string]*int{
+				{"a": pint(1), "b": pint(2), "z": pint(100)},
+				{"b": pint(22), "c": pint(3), "a": nil},
+				{"c": pint(33), "d": pint(4), "a": pint(111), "b": nil, "deleted": nil},
+			},
+			wantMap: map[string]int{"z": 100, "c": 33, "d": 4, "a": 111},
+		},
+		"CommitMultipleLayersDeleteAndAddBack": {
+			patchLayers: []map[string]*int{
+				{"x": pint(10)},
+				{"x": nil},
+				{"x": pint(30)},
+				{"y": pint(40), "x": nil}},
+			wantMap: map[string]int{"y": 40},
+		},
+		"CommitEmptyLayersBetweenDataLayers": {
+			patchLayers: []map[string]*int{{"a": pint(1)}, {}, {"b": pint(2)}, {}, {"c": pint(3)}},
+			wantMap:     map[string]int{"a": 1, "b": 2, "c": 3},
+		},
+	}
+
+	for testName, tc := range tests {
+		t.Run(testName, func(t *testing.T) {
+			ps := buildTestPatchSet(t, tc.patchLayers)
+			initialNumPatches := len(ps.patches)
+
+			if currentMap := ps.AsMap(); !maps.Equal(currentMap, tc.wantMap) {
+				t.Errorf("AsMap() before any commits mismatch: got %v, want %v", currentMap, tc.wantMap)
+			}
+
+			for i := 0; i < initialNumPatches-1; i++ {
+				expectedPatchesAfterCommit := len(ps.patches) - 1
+				ps.Commit()
+				if len(ps.patches) != expectedPatchesAfterCommit {
+					t.Errorf("After commit #%d, expected %d patches, got %d", i+1, expectedPatchesAfterCommit, len(ps.patches))
+				}
+				if currentMap := ps.AsMap(); !maps.Equal(currentMap, tc.wantMap) {
+					t.Errorf("AsMap() after commit #%d mismatch: got %v, want %v", i+1, currentMap, tc.wantMap)
+				}
+			}
+
+			if initialNumPatches > 0 && len(ps.patches) != 1 {
+				t.Errorf("Expected 1 patch after all commits, got %d", len(ps.patches))
+			} else if initialNumPatches == 0 && len(ps.patches) != 0 {
+				t.Errorf("Expected 0 patches after all commits, got %d", len(ps.patches))
+			}
+		})
+	}
+}
+
+func TestPatchSetRevert(t *testing.T) {
+	tests := map[string]struct {
+		patchLayers         []map[string]*int
+		wantInitialMap      map[string]int
+		wantMapsAfterRevert []map[string]int
```

Comment:
nit: I'd expect this map to be 1 element shorter than patch layers instead of having 2 first elements duplicated. Even if we want to test reverting a single layer, we can use `wantMapsAfterRevert[0]` for that in the test code.